File: rust/examples/list_polygons.py
"""Imprime area total por capa, por celda. Espejo de list_polygons.rs."""
import sys
from collections import defaultdict

import gdstk


def main() -> None:
    if len(sys.argv) < 2:
        print("usage: list_polygons.py <file.gds>", file=sys.stderr)
        sys.exit(2)

    lib = gdstk.read_gds(sys.argv[1])
    for cell in lib.cells:
        by_layer: dict[int, float] = defaultdict(float)
        for poly in cell.polygons:
            by_layer[poly.layer] += poly.area()
        print(f"Cell '{cell.name}':")
        for layer in sorted(by_layer):
            print(f"  Layer {layer}: {by_layer[layer]:.2f}")


if __name__ == "__main__":
    main()

File: rust/examples/list_polygons.rs
//! Imprime área total por capa, por celda. Fase 2 test de paridad con Python.
//!
//! Usage: list_polygons <file.gds>

use std::collections::BTreeMap;
use std::env;

use gdstk_rs::Library;

fn main() {
    let mut args = env::args();
    let program = args.next().unwrap_or_else(|| "list_polygons".into());
    let Some(path) = args.next() else {
        eprintln!("usage: {program} <file.gds>");
        std::process::exit(2);
    };

    let lib = Library::open(&path);

    for cell in lib.cells() {
        let mut by_layer: BTreeMap<u32, f64> = BTreeMap::new();
        for poly in cell.polygons() {
            *by_layer.entry(poly.layer()).or_insert(0.0) += poly.area();
        }
        println!("Cell '{}':", cell.name());
        for (layer, area) in &by_layer {
            println!("  Layer {layer}: {area:.2}");
        }
    }
}
